### ISSUE TITLE
Don't set SIGTERM handler on import

### ISSUE THREAD

**Issue by gasteigerjo** [ray-project/ray#17745]
### What is the problem?

When importing Ray, any previous SIGTERM handlers are disabled due to Ray overwriting the SIGTERM handler outside of any function here: https://github.com/ray-project/ray/blob/7a0b8982f3195182f17aca54232165441689c6f5/python/ray/worker.py#L1036

A simple import thus can have large unintended side effects. For example, in our repository this caused submitit (the Slurm management system we use) to no longer properly handle job preempting, even though no functionality of Ray was actually used in those runs. Due to submitit design reasons the Ray import actually happens after submitit sets the signal handler.

Imho an import shouldn't have such side effects.

*Ray version and other system information (Python version, TensorFlow version, OS):* Ray 1.4.0, Python 3.8.10, Ubuntu 20.04

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):

```
import signal
import time
signal.signal(signal.SIGTERM, signal.SIG_IGN)

from ray import tune

time.sleep(100)
# SIGTERM signal should be ignored now, but isn't due to side effects from importing ray.
# Check this via `kill -SIGTERM [pid]`
```

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

**Comment by mustafaKus02** [comment]
Hi, Is this solved?

**Comment by rkooo567** [comment]
Hey @gasteigerjo can you tell us if we defer to set sigterm after ray.init(), would that be acceptable, or is it the same?

**Comment by gasteigerjo** [comment]
That would solve it in the use case of the Open Catalyst Project, since `init` would only be called in the cases where Ray is actually used.

If that behavior is documented for the init function, i.e. in its docstring, then I think this should be acceptable.

**Comment by gerritdeen** [comment]
This specific problem broke my prefect deployments. A signal handler like the sigterm_handler in the line (in worker.py):

|ray._private.utils.set_sigterm_handler(sigterm_handler)

can only be set from the main thread of the main interpreter. So if you have any programs that import ray in a seperate thread it's possible / likely that the signal handling of the parent process gets broken. In such a case you have to ensure that the parent process does the signal handling correctly ensuring a proper cleanup on exit or you have to ensure it yourself. But tbh, I don't think many people will be calling ray.init from a non-main thread, so as long as the main thread has a sigterm_handler, we should be good. 

I can only speak for prefect, but I imagine other runners / executors use threading to keep track of jobs, so I think if we just add a check if we're indeed in the main thread, it's probably fine, like so:

import sys
import signal
import threading
import logging

logger = logging.getLogger(__name__)
logger.setLevel("INFO")

def sigterm_handler(signal, frame):
    logger.info("Received SIGTERM signal. Exiting gracefully...")
    # Perform cleanup or other necessary actions here
    sys.exit(0)

def set_conditional_sigterm_handler():
    current_thread = threading.current_thread()
    main_thread = threading.main_thread()

    if current_thread == main_thread:
        # Set SIGTERM handler in the main thread
        signal.signal(signal.SIGTERM, sigterm_handler)
        logger.info("SIGTERM handler set in the main thread.")
    else:
        logger.info("Not in the main thread. SIGTERM handler not set.")

def worker_thread():
    set_conditional_sigterm_handler()

# Call the function from the main thread
set_conditional_sigterm_handler()

# Create a worker thread and call the function from the threa
thread = threading.Thread(target=worker_thread)
thread.start()
thread.join()

The first call results in 15:22:34.074 | INFO    | wpp - SIGTERM handler set in the main thread.
The second call results in 15:22:34.076 | INFO    | wpp - Not in the main thread. SIGTERM handler not set.

**Comment by pranaymodukuru** [comment]
Any update on the resolution of this issue?

**Comment by gerritdeen** [comment]
I did a bit more digging and I don't think it has to do with the signal handling, I tested a version that only sets the ray signal handler if it's in the main thread, but that didn't solve it.

When I looked into a docker container that didn't die, I saw that it was waiting for resources using strace, something like this:

futex(0x7f58f8000990, FUTEX_WAIT, 2, NULL

This didn't happen when I didn't use ray with prefect and it only happened when the docker container was trying to run jobs that involved the concurrent task runner. This led me to the hypothesis that ray must be doing something involving threading which somehow causes prefect not being able to run concurrent tasks. 

For now we just avoid this scenario where jobs use both ray and the concurrent task runner and only import ray when we actually use the ray task runner. In the future we may want to run all our jobs on ray, so we may make
